main.go

```diff
@@ -1,112 +1,148 @@
-		logrus.Fatalf("Error building kubeconfig: %s", err.Error())
+	reconciler := &controller.GitJobReconciler{
+		Client:    mgr.GetClient(),
+		Scheme:    mgr.GetScheme(),
+		Image:     flags.image,
+		GitPoller: poll.NewHandler(mgr.GetClient()),
+		Log:       ctrl.Log.WithName("gitjob-reconciler"),
 	}
-	cfg.RateLimiter = ratelimit.None
 
-	cont := types.BuildContext(c.String("namespace"), cfg)
-	if err := cont.Start(ctx); err != nil {
-		logrus.Fatal(err)
-	}
-	cont.Image = c.String("gitjob-image")
+	group := errgroup.Group{}
+	group.Go(func() error {
+		return startWebhook(ctx, namespace, flags.listen, mgr.GetClient(), mgr.GetCache())
+	})
+	group.Go(func() error {
+		setupLog.Info("starting manager")
+		if err = reconciler.SetupWithManager(mgr); err != nil {
+			return err
+		}
 
-	go func() {
-		leader.RunOrDie(ctx, c.String("namespace"), "gitjob", cont.K8s, func(ctx context.Context) {
-			controller.Register(ctx, cont)
-			runtime.Must(cont.Start(ctx))
-			<-ctx.Done()
-		})
-	}()
+		return mgr.Start(ctx)
+	})
+
+	return group.Wait()
+}
+
+func bindFlags() flags {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var image string
+	var listen string
+	var debug bool
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8081", "The address the metric endpoint binds to.")
+	flag.StringVar(&image, "gitjob-image", "rancher/gitjob:dev", "The gitjob image that will be used in the generated job.")
+	flag.StringVar(&listen, "listen", ":8080", "The port the webhook listens.")
+	flag.BoolVar(&enableLeaderElection, "leader-elect", true,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&debug, "debug", false, "debug mode.")
+	opts := zap.Options{
+		Development: debug,
+	}
+	opts.BindFlags(flag.CommandLine)
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+	flag.Parse()
 
-	logrus.Info("Setting up webhook listener")
-	addr := c.String("listen")
+	return flags{
+		metricsAddr:          metricsAddr,
+		enableLeaderElection: enableLeaderElection,
+		image:                image,
+		listen:               listen,
+		debug:                debug,
+	}
+}
 
+func startWebhook(ctx context.Context, namespace string, addr string, client client.Client, cacheClient cache.Cache) error {
+	setupLog.Info("Setting up webhook listener")
+	handler, err := webhook.HandleHooks(ctx, namespace, client, cacheClient)
+	if err != nil {
+		return fmt.Errorf("webhook handler can't be created: %w", err)
+	}
 	server := &http.Server{
 		Addr:    addr,
-		Handler: webhook.HandleHooks(ctx, cont),
+		Handler: handler,
 		// According to https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
 		ReadHeaderTimeout: 5 * time.Second,
 		WriteTimeout:      10 * time.Second,
 	}
 
-	if err := server.ListenAndServe(); err != nil {
-		logrus.Fatalf("Failed to listen on %s: %v", addr, err)
+	go func() {
+		<-ctx.Done()
+		server.Shutdown(context.Background())
```

Comment:
I would add a short shutdown timeout instead of `context.Background()`, e.g.:
```suggestion
		shutdownCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()
		server.Shutdown(shutdownCtx)
```